fl4health/utils/load_data.py

```diff
@@ -1,85 +1,138 @@
+import random
 from logging import INFO
 from pathlib import Path
-from typing import Callable, Dict, Optional, Tuple, Union
+from typing import Callable, Dict, Optional, Tuple
 
+import torch
 import torchvision.transforms as transforms
 from flwr.common.logger import log
 from torch.utils.data import DataLoader
+from torchvision.datasets import CIFAR10, MNIST
 
-from fl4health.utils.dataset import BaseDataset, Cifar10Dataset, MnistDataset
+from fl4health.utils.dataset import TensorDataset
 from fl4health.utils.dataset_converter import DatasetConverter
+from fl4health.utils.random import set_all_random_seeds
 from fl4health.utils.sampler import LabelBasedSampler
 
 
+def split_data_and_targets(
+    data: torch.Tensor, targets: torch.Tensor, validation_proportion: float = 0.02
+) -> Tuple[torch.Tensor, torch.Tensor, torch.Tensor, torch.Tensor]:
+
+    total_size = data.shape[0]
+    train_size = int(total_size * (1 - validation_proportion))
+    train_indices = random.sample(range(total_size), train_size)
+    val_indices = [i for i in range(total_size) if i not in train_indices]
+    train_data, train_targets = data[train_indices, ...], targets[train_indices, ...]
+    val_data, val_targets = data[val_indices, ...], targets[val_indices, ...]
+    return train_data, train_targets, val_data, val_targets
+
+
+def get_mnist_data_and_target_tensors(data_dir: Path, train: bool) -> Tuple[torch.Tensor, torch.Tensor]:
+    mnist_dataset = MNIST(data_dir, train=train, download=True)
+    data = torch.Tensor(mnist_dataset.data)
+    targets = torch.Tensor(mnist_dataset.targets).long()
+    return data, targets
+
+
+def get_train_and_val_mnist_datasets(
+    data_dir: Path,
+    transform: Optional[Callable] = None,
+    target_transform: Optional[Callable] = None,
+    validation_proportion: float = 0.2,
+) -> Tuple[TensorDataset, TensorDataset]:
+    data, targets = get_mnist_data_and_target_tensors(data_dir, True)
+
+    train_data, train_targets, val_data, val_targets = split_data_and_targets(data, targets, validation_proportion)
+
+    training_set = TensorDataset(train_data, train_targets, transform=transform, target_transform=target_transform)
+    validation_set = TensorDataset(val_data, val_targets, transform=transform, target_transform=target_transform)
+    return training_set, validation_set
+
+
 def load_mnist_data(
     data_dir: Path,
     batch_size: int,
     sampler: Optional[LabelBasedSampler] = None,
-    transform: Union[None, Callable] = None,
-    target_transform: Union[None, Callable] = None,
+    transform: Optional[Callable] = None,
+    target_transform: Optional[Callable] = None,
     dataset_converter: Optional[DatasetConverter] = None,
+    validation_proportion: float = 0.2,
+    seed: int = 42,
 ) -> Tuple[DataLoader, DataLoader, Dict[str, int]]:
     """Load MNIST Dataset (training and validation set)."""
     log(INFO, f"Data directory: {str(data_dir)}")
+
+    set_all_random_seeds(seed)
```

Comment:
I think, but I'm not 100% sure, that this will actually set a global seed, not just for the loading and preparing of the data. It will also override any other seeds that someone has set (i.e. in other parts of their flow). So rather, I think we strictly want to seed locally here. That is, the seed should be optional and none by default, so that the user **has to** explicitly set the seed again if they want it. If it's set, we only set the seeds for local generators that are necessary for reproducibility of the things like splitting and sampling.